transaction/core/src/memo.rs

```diff
@@ -0,0 +1,153 @@
+//!   functionality can be used to assist.
+
+use aes::{
+    cipher::{FromBlockCipher, StreamCipher},
+    Aes256, Aes256Ctr, NewBlockCipher,
+};
+use core::convert::{TryFrom, TryInto};
+use displaydoc::Display;
+use generic_array::{
+    sequence::Split,
+    typenum::{U32, U34, U48},
+    GenericArray,
+};
+use hkdf::Hkdf;
+use mc_crypto_keys::{CompressedRistrettoPublic, RistrettoPublic};
+use serde::{Deserialize, Serialize};
+use sha2::Sha512;
+
+/// A plaintext memo payload, with accessors to easily access the memo type
+/// bytes and memo data bytes. High-level memo objects should be convertible
+/// to MemoPayload. Deserialization, across all high-level memo types, is
+/// done in mc-memo crate.
+///
+/// Note that a memo payload may be invalid / uninterpretable, or refer to new
+/// memo types that have been introduced at a later date.
+#[derive(Clone, Deserialize, Default, Eq, Ord, PartialEq, PartialOrd, Serialize, Debug)]
+pub struct MemoPayload(GenericArray<u8, U34>);
+
+impl MemoPayload {
+    /// Create a new memo payload from given type bytes and data bytes
+    pub fn new(memo_type: [u8; 2], memo_data: [u8; 32]) -> Self {
+        let mut result = Self::default();
+        result.0[0..2].copy_from_slice(&memo_type);
+        result.0[2..34].copy_from_slice(&memo_data);
+        result
+    }
+
+    /// Get the memo type bytes (two bytes)
+    pub fn get_memo_type(&self) -> &[u8; 2] {
+        self.0.as_slice()[0..2].try_into().unwrap()
+    }
+
+    /// Get the memo data bytes (thirty-two bytes)
+    pub fn get_memo_data(&self) -> &[u8; 32] {
+        self.0.as_slice()[2..34].try_into().unwrap()
+    }
+
+    /// Encrypt this memo payload using a given shared-secret, consuming it and
+    /// returning underlying buffer.
+    ///
+    /// The shared-secret is expected to be the TxOut shared secret associated
+    /// to the memo.
+    pub fn encrypt(mut self, shared_secret: &RistrettoPublic) -> GenericArray<u8, U34> {
+        self.apply_keystream(&shared_secret);
+        self.0
+    }
+
+    /// Try to decrypt given e_memo bytes using a shared secret.
+    ///
+    /// Note: The results of this call are unauthenticated.
+    pub fn try_decrypt(src: &[u8], shared_secret: &RistrettoPublic) -> Result<Self, LengthError> {
+        let mut result = Self::try_from(src)?;
+        result.apply_keystream(&shared_secret);
+        Ok(result)
+    }
+
+    // Apply AES256 keystream to internal buffer.
+    // This is not a user-facing API, since from the user's point of view this
+    // object always represents decrypted bytes.
+    //
+    // The argument is supposed to be the TxOut shared secret associated to the
+    // memo.
+    fn apply_keystream(&mut self, shared_secret: &RistrettoPublic) {
+        // Use HKDF-SHA512 to produce an AES key and AES nonce
+        let shared_secret = CompressedRistrettoPublic::from(shared_secret);
+        let kdf = Hkdf::<Sha512>::new(Some(b"mc-memo-okm"), shared_secret.as_ref());
```

Comment:
whats `okm`?

Reply:
okm is output key material

the rfc for hkdf describes "input key material", "salt", "info-string", and then the output is called "output key material"

i could spell this out, it's probably better

Reply:
Thanks for explaining. I'm fine with `okm` + a comment